bot/seasons/easter/egg_hunt/cog.py

```diff
@@ -0,0 +1,511 @@
+            self.timeout_task = None
+
+        await asyncio.sleep(seconds)
+
+        bot.remove_listener(self.collect_reacts, name="on_reaction_add")
+
+        with contextlib.suppress(discord.Forbidden):
+            await self.message.clear_reactions()
+
+        if self.first:
+            self.finalise_score()
+
+    def is_valid_react(self, reaction: discord.Reaction, user: discord.Member) -> bool:
+        """Validates a reaction event was meant for this session."""
+
+        if user.bot:
+            return False
+        if reaction.message.id != self.message.id:
+            return False
+        if reaction.emoji != self.egg:
+            return False
+        return True
+
+    async def collect_reacts(self, reaction: discord.Reaction, user: discord.Member):
+        """Handles emitted reaction_add events via listener."""
+
+        if not self.is_valid_react(reaction, user):
+            return
+
+        team = get_team_role(user)
+        if not team:
+            log.debug(f"Assigning a team for {user}.")
+            user = await assign_team(user)
+
+        if not self.first:
+            log.debug(f"{user} was first to react to egg on {self.message.id}.")
+            self.first = user
+            await self.start_timeout()
+        else:
+            self.users.append(user)
+
+    async def start(self):
+        """Starts the egg drop session."""
+
+        log.debug(f"EggHunt session started for message {self.message.id}.")
+        bot.add_listener(self.collect_reacts, name="on_reaction_add")
+        with contextlib.suppress(discord.Forbidden):
+            await self.message.add_reaction(self.egg)
+        self.timeout_task = asyncio.create_task(self.start_timeout(300))
+
+
+class SuperEggMessage(EggMessage):
+    """Handles a super egg session."""
+
+    def __init__(self, message: discord.Message, egg: discord.Emoji, window: int):
+        super().__init__(message, egg)
+        self.window = window
+
+    async def finalise_score(self):
+        """Sums and actions scoring for this super egg session."""
+        try:
+            message = await self.message.channel.get_message(self.message.id)
+        except discord.NotFound:
+            return
+
+        count = 0
+        white = 0
+        blurple = 0
+        react_users = []
+        for reaction in message.reactions:
+            if reaction.emoji == self.egg:
+                react_users = await reaction.users().flatten()
+                for user in react_users:
+                    team = get_team_role(user)
+                    if team == Roles.white:
+                        white += 1
+                    elif team == Roles.blurple:
+                        blurple += 1
+                count = reaction.count - 1
+                break
+
+        score = 50 if self.egg == Emoji.egg_gold else 100
+        if white == blurple:
+            log.debug("Tied SuperEgg Result.")
+            team = None
+            score = score / 2
```

Comment:
I'd prefer to see `score /= 2` here, but won't delay approval over it.

Reply:
Had no idea this was a thing. Thanks.